mephisto/core/task_launcher.py

```diff
@@ -45,63 +53,87 @@ def __init__(
+        self.finished_generators: bool = False
+        self.unlaunched_units_access_condition = threading.Condition()
+        if isinstance(self.assignment_data_iterable, types.GeneratorType):
+            self.generator_type = GeneratorType.ASSIGNMENT
+        else:
+            self.generator_type = GeneratorType.NONE
         run_dir = task_run.get_run_dir()
         os.makedirs(run_dir, exist_ok=True)
 
-    def create_assignments(self) -> None:
-        """
-        Create an assignment and associated units for any data
-        currently in the assignment config
-        """
+        logger.debug(f"type of assignment data: {type(self.assignment_data_iterable)}")
+
+    def _create_single_assignment(self, assignment_data) -> None:
+        """ Create a single assignment in the database using its read assignment_data """
         task_run = self.task_run
         task_config = task_run.get_task_config()
-        for data in self.assignment_data_list:
-            assignment_id = self.db.new_assignment(
+        assignment_id = self.db.new_assignment(
+            task_run.task_id,
+            task_run.db_id,
+            task_run.requester_id,
+            task_run.task_type,
+            task_run.provider_type,
+            task_run.sandbox,
+        )
+        assignment = Assignment(self.db, assignment_id)
+        assignment.write_assignment_data(assignment_data)
+        self.assignments.append(assignment)
+        unit_count = len(assignment_data["unit_data"])
+        for unit_idx in range(unit_count):
+            unit_id = self.db.new_unit(
                 task_run.task_id,
                 task_run.db_id,
                 task_run.requester_id,
-                task_run.task_type,
+                assignment_id,
+                unit_idx,
+                task_config.task_reward,
                 task_run.provider_type,
+                task_run.task_type,
                 task_run.sandbox,
             )
-            assignment = Assignment(self.db, assignment_id)
-            assignment.write_assignment_data(data)
-            self.assignments.append(assignment)
-            unit_count = len(data["unit_data"])
-            for unit_idx in range(unit_count):
-                unit_id = self.db.new_unit(
-                    task_run.task_id,
-                    task_run.db_id,
-                    task_run.requester_id,
-                    assignment_id,
-                    unit_idx,
-                    task_config.task_reward,
-                    task_run.provider_type,
-                    task_run.task_type,
-                    task_run.sandbox,
-                )
-                self.units.append(Unit(self.db, unit_id))
+            self.units.append(Unit(self.db, unit_id))
+            with self.unlaunched_units_access_condition:
                 self.unlaunched_units[unit_id] = Unit(self.db, unit_id)
-                self.keep_launching = True
+            self.keep_launching_units = True
```

Comment:
This should only be set once when `launch_assignments` is first called, or else an errant call here could lead to a race condition against `expire_units` where the launching continues.

Reply:
I'll modify it but can we discuss how the race condition can happen over the meeting?